From cf382341dc452bb7539406a1d584e60b5a4010df Mon Sep 17 00:00:00 2001
From: tanshanshan <tanshanshan@eswincomputing.com>
Date: Wed, 26 Mar 2025 15:29:35 +0800
Subject: [PATCH 476/652] fix:fix pacmsg bug

Changelogs:
1.add boundary handling code to fix first_seq calculation error.
2.add spin lock to resolve handle_page_fault bug caused
  by concurrency.

Signed-off-by: tanshanshan <tanshanshan@eswincomputing.com>
Reviewed-by: Zhou Liang <zhouliang1@eswincomputing.com>
Reviewed-by: Joey Zheng <zhengyu@eswincomputing.com>
Tested-by: Guichao Zhou <zhouguichao@eswincomputing.com>
---
 kernel/printk/printk.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/kernel/printk/printk.c b/kernel/printk/printk.c
index 9d25640e85250..c10dc4bc3e65a 100644
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -84,6 +84,8 @@ struct paclog_iter {
 	u64 log_first_seq;
 	u64 log_next_seq;
 };
+
+static DEFINE_SPINLOCK(shm_lock);
 static bool shm_ready_flag;
 static u64 reserved_next_seq;
 static void __iomem *shm_va;
@@ -2278,6 +2280,20 @@ static int reserve_log_to_share_memory(char *buffer, size_t line_len)
 	memset(&new_info, 0, sizeof(struct paclog_info));
 
 	if (next_idx + offset >= PACMSG_MAX_OFFSET) {
+		while (first_seq < next_seq) {
+			memcpy(&old_info, (void *)(pacmsg_buf + first_idx),
+			       sizeof(struct paclog_info));
+
+			if (old_info.len) {
+				if (first_idx >= next_idx) {
+					first_idx += old_info.len;
+					first_seq++;
+				} else
+					break;
+			} else
+				break;
+		}
+
 		memcpy((void *)(pacmsg_buf + next_idx), &new_info,
 		       sizeof(struct paclog_info));
 		next_idx = 0;
@@ -2325,6 +2341,9 @@ static int reserve_record_to_share_memory(void)
 	struct printk_record r;
 	char *text;
 	u64 seq;
+	unsigned long flags;
+
+	spin_lock_irqsave(&shm_lock, flags);
 
 	text = kmalloc(PRINTK_MESSAGE_MAX, GFP_ATOMIC);
 	if (!text)
@@ -2343,6 +2362,8 @@ static int reserve_record_to_share_memory(void)
 	}
 
 	kfree(text);
+
+	spin_unlock_irqrestore(&shm_lock, flags);
 	return 0;
 }
 #endif
-- 
2.53.0

